avocado/plugins/assets.py

```diff
@@ -0,0 +1,179 @@
+
+class FetchAssets(Test):
+    def test_fetch_assets(self):
+{fetch_assets}
+"""
+
+
+class FetchAssetHandler(ast.NodeVisitor):
+
+    def __init__(self, file_name):
+        self.file_name = file_name
+        # Right now, the class handles just 'fetch_assets' call.
+        # In the future, it can be extended to search for other call patterns.
+        self.pattern = 'fetch_asset'
+        self.calls = []
+        self.result = None
+
+        # create Abstract Syntax Tree from test source file
+        with open(self.file_name) as source_file:
+            self.tree = ast.parse(source_file.read(), self.file_name)
+
+        # build calls list
+        self.visit(self.tree)
+    # __init__()
+
+    def _build_call(self, node):
+        """
+        Build the call string from ast nodes.
+        :param node: Call root node
+        :type node: ast.Call
+        :return: Call prepared to be executed into a test class
+        :rtype: str
+        """
+        # we handle code indentation
+        call = "        self." + node.func.attr + "("
+        for a in node.args:
+            if isinstance(a, ast.Name):
+                return None
+            elif isinstance(a, ast.Str):
+                call = call + '"' + ast.literal_eval(a) + '", '
+        for k in node.keywords:
+            call = call + k.arg + "="
+            if isinstance(k.value, ast.Name):
+                return None
+            elif isinstance(k.value, ast.Str):
+                call = call + '"' + ast.literal_eval(k.value) + '", '
+        # remove last comma and add parenthesis
+        call = call[:-2] + ")\n"
+
+        return call
+    # _build_call()
+
+    def visit_Call(self, node):
+        """
+        Visit calls and build call string if call matches the pattern
+        :param node: (Sub)Tree root to start the parse
+        :type node: ast.*
+        """
+        if isinstance(node.func, ast.Attribute):
+            if self.pattern in node.func.attr:
+                call = self._build_call(node)
+                if call is not None:
+                    self.calls.append(call)
+    # visit_call()
+
+    def fetcher(self):
+        """
+        Build an intrumented avocado test with all fetch_asset calls and
+        execute it.
+        """
+        # build raw code from list of calls, if available
+        if len(self.calls) > 0:
+            rcalls = ""
+            for call in self.calls:
+                rcalls = rcalls + call
+        else:
+            LOG_UI.debug('No supported assets found to fetch from %s'
+                         % self.file_name)
+            return
+
+        # temporary test source file
+        fetch_test = tempfile.NamedTemporaryFile(suffix=".py")
+        fetch_test.write(TEST_TEMPLATE.format(fetch_assets=rcalls).encode())
+        fetch_test.flush()
+
+        # run teporary test
+        LOG_UI.debug('Fetching assets from %s' % self.file_name)
+        self.result = process.run("avocado run --json - %s" % fetch_test.name,
```

Comment:
And now about this, I don't really like the idea of creating a test file just to fetch the assets. Not sure what am I missing here, but wouldn't it be safer and easier to simply use:

```
cache_dirs = data_dir.get_cache_dirs()
err = []
for fetch_item in fetch_assets:
    try:
        expire = fetch_item.kwargs.pop('expire', None)
        # 5th positional argument is 6th argument in asset.Assets
        if len(fetch_item.args) == 5:
            expire = fetch_item.args[4]
            fetch_item.args = fetch_item.args[:4]
        elif len(fetch_item.args) > 5:
            err.append(...)
            continue
        if expire is not None:
            expire = data_structures.time_to_seconds(str(expire))
        asset.Asset(*fetch_item.args, **fetch_item.kwargs, cache_dirs=cache_dirs, expire=expire)
    except ...:
        err.append(...)
```

... only after preparing this I noticed the `metadata` new argument in `asset.Asset`, it might need to also be treated. Although it's not supported by `self.fetch_asset` so either we should add it there as well, or not to care about it at all.

Reply:
^^ btw this was written in a hurry and you might need to improve it (checking for duplicate args...), but it should serve as an inspiration.

Reply:
... and I forgot to mention that the `fetch_assets` is a list of arguments collected in `_build_call` (stored in `self.calls`)

Reply:
also note that by calling it directly you'd actually reflect potential `avocado --config ... assets fetch ...` changes. By calling another `avocado` process you might end-up fetching the assets to a different location(s).

Reply:
The main objective here was to make use of all logging structure avocado command gives.

> also note that by calling it directly you'd actually reflect potential `avocado --config ... assets fetch ...` changes. By calling another `avocado` process you might end-up fetching the assets to a different location(s).

When considering this use case, the current implementation does not fit it, so I will think about other ways to implement that, taking into consideration your suggestions.

Reply:
Well you won't lose the logging messages by directly calling the `asset.Asset`. Even better you can tweak the verbosity to show more or less info...

Reply:
Btw this is my strongest point here, I don't like the idea of generating a test just to fetch the assets. I mean I won't block this if someone else supports it, but using API directly is IMO much nicer.